crates/clvm-utils/src/tree_hash.rs

```diff
@@ -307,6 +308,79 @@ pub fn tree_hash_cached(a: &Allocator, node: NodePtr, cache: &mut TreeCache) ->
     hashes[0]
 }
 
+fn subtract_cost(cost_left: &mut Cost, subtract: Cost) -> Result<(), usize> {
+    if subtract > *cost_left {
+        Err(0)
+    } else {
+        *cost_left -= subtract;
+        Ok(())
+    }
+}
+
+pub fn tree_hash_cached_costed(
+    a: &Allocator,
+    node: NodePtr,
+    cache: &mut TreeCache,
+    cost_left: &mut u64,
+    cost_per_call: u32,
+    cost_per_byte: u32,
+) -> Result<TreeHash, usize> {
+    cache.visit_tree(a, node);
+
+    let mut hashes = Vec::new();
+    let mut ops = vec![TreeOp::SExp(node)];
+
+    while let Some(op) = ops.pop() {
+        subtract_cost(cost_left, cost_per_call.into())?;
+        match op {
+            TreeOp::SExp(node) => match a.node(node) {
+                NodeVisitor::Buffer(bytes) => {
+                    subtract_cost(cost_left, (cost_per_byte * bytes.len() as u32).into())?;
+                    let hash = tree_hash_atom(bytes);
+                    hashes.push(hash);
+                }
+                NodeVisitor::U32(val) => {
+                    subtract_cost(cost_left, (cost_per_byte * 4).into())?; // 4 bytes in u32
```

Comment:
this is not necessarily 4 bytes. There should be a unit test to cover this case, with short atoms. 1, 2, 3, 4 and 5 bytes.

In `clvm_rs` there's a function `len_for_value()` which can tell you how many bytes. But you can also ask the `Allocator`:
```
a.atom_len(node)
```